include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh

```diff
@@ -962,6 +962,106 @@ class open_addressing_ref_impl {
     }
   }
 
+  /**
+   * @brief Executes a callback on every element in the container with key equivalent to the probe
+   * key.
+   *
+   * @note Passes an un-incrementable input iterator to the element whose key is equivalent to
+   * `key` to the callback.
+   *
+   * @tparam ProbeKey Input type which is convertible to 'key_type'
+   + @tparam Callback Callback functor or lambda
+   *
+   * @param key The key to search for
+   * @param callback Function to call on every element found
+   */
+  template <class ProbeKey, class Callback>
+  __device__ void for_each(ProbeKey const& key, Callback callback) const noexcept
```

Comment:
```suggestion
  __device__ void for_each(ProbeKey const& key, Callback& callback) const noexcept
```

Unsure if this needs to be a mutable reference instead. Let's say we define a `count` functor as such:

```cpp
struct count_functor {
  std::size_t thread_count = 0;

  // counts the number of matching elements for this thread
  template <class InputIt>
  __device__ void operator()(InputIt) { thread_count++; }
};
```

And then call
```cpp
//...
auto thread_counter = count_functor{};
set.for_each(key, thread_counter);
auto const key_count = thread_counter.count;
//...
```

Then we want the functor to be taken as a mutable reference.

Reply:
pass by value is preferred.

The above example is a good example of a bad callback especially in a parallel context

Reply:
Is it best practice to pass a callback by-value? I'd have to skim some stackoverflow/cppreference pages to get familiar with the topic. With pass-by-value we lose the ability of giving the callback an internal state that can hold the result of the operation. How would we solve the above example with a callback passed by-value? Pass a pointer to `thread_count` to the callback?

Reply:
> Pass a pointer to thread_count to the callback?

Yes

I see a callable defining the operations to be performed on the output instead of being the output itself.